Adding parallel support for the LLVM backend.

zheng-xq opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53243 Adding parallel support for the LLVM backend.**

Differential Revision: [D26906509](https://our.internmc.facebook.com/intern/diff/D26906509)

zheng-xq committed:
Adding parallel support for the LLVM backend.

[ghstack-poisoned]

zheng-xq committed:
Update on "Adding parallel support for the LLVM backend."




[ghstack-poisoned]

zheng-xq committed:
Update on "Adding parallel support for the LLVM backend."




[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -229,6 +242,39 @@ class LLVMCodeGenImpl : public IRVisitor {
 };
+
+// TODO: more extensive testing.
+// TODO: benchmarking.
```
There are some benchmarks in the PR, do you plan to add something more? If so, it would be helpful to elaborate in the comment (and similarly in the previous comment about more tests).

ZolotukhinM commented on test/cpp/tensorexpr/test_llvm.cpp:
```diff
@@ -1610,6 +1610,133 @@ TEST(LLVM, RFactorVectorizedReduction) {
+        loop_list[i]->set_parallel();
+      }
+    }
+    Placeholder t1_buf(BufHandle(t1->buf()));
+    Placeholder t2_buf(BufHandle(t2->buf()));
+    Placeholder t3_buf(BufHandle(t3->buf()));
+    Placeholder t4_buf(BufHandle(t4->buf()));
```
You do not need to create placeholders for Tensors, they are only needed for external buffers, for which we don't have `Compute`.

ZolotukhinM commented on test/cpp/tensorexpr/test_llvm.cpp:
```diff
@@ -1610,6 +1610,133 @@ TEST(LLVM, RFactorVectorizedReduction) {
+    Placeholder t4_buf(BufHandle(t4->buf()));
+    loop_nest.prepareForCodegen();
+    Stmt* stmt = loop_nest.root_stmt();
+    LLVMCodeGen cg(stmt, {t4_buf});
```
You could pass `t4` directly here:
```
LLVMCodeGen cg(stmt, {t4});
```
I noticed that there are other tests doing a similar thing, but I believe that's accidental or some leftovers.

ZolotukhinM commented on test/cpp/tensorexpr/test_llvm.cpp:
```diff
@@ -1610,6 +1610,133 @@ TEST(LLVM, RFactorVectorizedReduction) {
+    }
+    loop_nest.prepareForCodegen();
+    Stmt* stmt = loop_nest.root_stmt();
+    Placeholder f_buf(BufHandle(tensor->buf()));
+    LLVMCodeGen cg(stmt, {f_buf});
```
Placeholder is not needed here, you could pass `tensor` directly to the codegen constructor:
```
LLVMCodeGen cg(stmt, {tensor});
```

ZolotukhinM commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -447,6 +494,7 @@ class LLVMIntrinsicsExpander : public GenericIntrinsicsExpander {
 
 void LLVMCodeGenImpl::emitKernel(
     Stmt* stmt,
+
```
Nit: unnecessary whitespace change.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -1109,7 +1180,133 @@ void LLVMCodeGenImpl::visit(const Load* v) {
+  v->body()->accept(&var_finder);
+  const auto& vars = var_finder.vars();
+  for (auto& var : vars) {
+    llvm::Value* value = varToValue(var);
+    if (value) {
```
Nit: can do in one line :)

ZolotukhinM reviewed (approved):
This is awesome! I have just a couple of small nits, but since the code is quite complicated (because it solves quite a complicated problem!) I'd recommend to wait for someone else to take one more look at it.

zheng-xq committed:
Update on "Adding parallel support for the LLVM backend."




[ghstack-poisoned]

zheng-xq commented:
All comments addressed. PTAL. Thanks!

navahgar reviewed (approved):
LGTM

zheng-xq committed:
Update on "Adding parallel support for the LLVM backend."


Differential Revision: [D26906509](https://our.internmc.facebook.com/intern/diff/D26906509)

[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -139,6 +170,8 @@ class LLVMCodeGenImpl : public IRVisitor {
 #undef LLVM_TYPE_DECLARE
   llvm::Type* Int8PtrTy_;
+  llvm::Type* VoidTy_;
+  llvm::Type* VoidPtrTy_;
```
Looks like this (`VoidPtrTy_`) is unused?

bertmaher commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -229,6 +270,37 @@ class LLVMCodeGenImpl : public IRVisitor {
+      USE_TRIGGER(llvm_codegen_parallel_dispatched);
+      std::unique_lock<std::mutex> l(m);
+      remaining_tasks--;
+      while (remaining_tasks != 0) {
+        cv.wait(l);
+      }
```
I usually like to use the predicate version of `wait` e.g. `cv.wait(l, [&] { remaining == 0 })` but this is fine too

bertmaher commented on torch/csrc/jit/tensorexpr/llvm_jit.h:
```diff
@@ -16,6 +16,13 @@ namespace torch {
 namespace tensorexpr {
 
+void DispatchParallel(int8_t* func, int start, int stop, int8_t* packed_data);
+typedef void (*TypeDispatchParallel)(
+    int8_t* func,
+    int start,
+    int stop,
+    int8_t* packed_data);
+
```
TypeDispatchParallel looks unused, remove if so.

bertmaher commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -1136,6 +1357,7 @@ void LLVMCodeGenImpl::visit(const For* v) {
   auto exit = llvm::BasicBlock::Create(getContext(), "exit", fn_);
 
   // Create the stop condition.
+  irb_.SetInsertPoint(condBlock);
```
Is this needed?  I'm a little surprised since we just do `irb_.SetInsertPoint(condBlock);` a few lines ago (1344) and I don't see anything that would invalidate that...

bertmaher reviewed (approved):
Very cool!

ZolotukhinM commented:
Looks like a number of approves and reaccs record! 👍

zheng-xq committed:
Update on "Adding parallel support for the LLVM backend."


Differential Revision: [D26906509](https://our.internmc.facebook.com/intern/diff/D26906509)

[ghstack-poisoned]
